README.markdown
`acts_as_solr Rails plugin`
======
This plugin adds full text search capabilities and many other nifty features from Apache's [Solr](http://lucene.apache.org/solr/) to any Rails model.
It was based on the first draft by Erik Hatcher.

Current Release
======
The current stable release is v1.0 and was released on 20-05-2009.

Changes
======
Please refer to the CHANGE_LOG

Installation
======

script/plugin install git://github.com/mauricio/acts_as_solr.git

Requirements
=====
* Java Runtime Environment(JRE) 1.5 aka 5.0 or higher [http://java.sun.com/](http://java.sun.com/)

Features
======

* Fully compatible with will_paginate view helpers and :page/:per_page options
* Using Solr 1.3.0

Things to be done
======

* Simplify configuration
* Examples using the "did you mean?" feature
* Upgrade and improve Jetty config

Mantainer
======
Maurício Linhares (mauricio dot linhares AT gmail dot com)

Original Authors
======
* Erik Hatcher
* Thiago Jackiw

Release Information
======
Released under the MIT license.

More info
======
[http://github.com/mauricio/acts_as_solr/tree/master](http://github.com/mauricio/acts_as_solr/tree/master)

Basic Usage
======
<pre><code>
# Just include the line below to any of your ActiveRecord models:

  class Book < ActiveRecord::Base
    acts_as_solr
  end

# With this acts_as_solr will index all string and text fields on your
# active_record model

# Or if you want, you can specify only the fields that should be indexed:

  class Book < ActiveRecord::Base
    acts_as_solr :fields => [:name, :author]
  end

# The "fields" that are going to be indexed don't need to be database fields,
# they can be simple methods on your model:

  class Book < ActiveRecord::Base

    belongs_to :author

    acts_as_solr :fields => [:name, :author_name]

    def author_name
      self.author.name if self.author
    end

  end

# And this is all there is to it, you don't need to get your hands dirty on
# Solr schema or document files, only if the default config doesn't suit your
# needs. After every save/destroy the Solr index for that object is going to be
# updated, so all you need to do is configure your active_record object as above
# start the Solr server (rake solr:start) and be done with it.

# If you're starting up from an already populated database, you'll have to build
# and index based on that data. To do this, start the Solr server (rake solr:start)
# and then call the reindex task (rake solr:rebuild_index), this will rebuild
# the index based on all objects currently stored on your database.

# Then to find instances of your model, just do:
  books = Book.find_by_solr(query)

# "query" is a string representing your query
# You can find out more about Lucene's query sintax here:
# http://lucene.apache.org/java/2_4_1/queryparsersyntax.html

# Finding using pagination:

  books = Book.find_by_solr( query, :page => 2, :per_page => 10 )

# The object returned by a find_by_solr call is an ActsAsSolr::SearchResults
# object where you can find out more about the response Solr sent back to you and
# the real active_record objects returned by your query.

# This object works just like any other Enumerable (as an Array) and you can
# call any Enumerable method on it. It's also completely compatible with 
# will_paginate view helpers.

  books = Book.find_by_solr( query, :page => 2, :per_page => 10 )
  books.each do |book| # this "book" variable is a real Book object from the database
    puts "#{book.name} - #{book.author_name} - #{book.year}"
  end

# Please see ActsAsSolr::ActsMethods for a complete info

</code></pre>

CHANGE_LOG
== CHANGE_LOG

=== 20-05-2009: Version 1.0

<b>NEW</b>:: No need to call "records" at the returned object from find_by_solr,
  now it's an Enumerable and behaves just like any other Array object.
<b>NEW</b>:: Upgraded Solr to 1.3.0
<b>NEW</b>:: Added support to will_paginate's style pagination by having 
  :page and :per_page options at the find_by_solr calls and also updated
  the pagination methods to work correctly with the latest will_paginate's
  view helpers.
<b>NEW</b>:: Added support for a :conditions option to find_by_solr calls to further filter the returned results.
<b>NEW</b>:: Added rake task to rebuild the indexes for all acts_as_solr enabled models

=== 06-18-2007: Version 0.9
<b>NEW</b>:: Added the option :scores when doing a search. If set to true this will return the score as a 'solr_score' attribute or each one of the instances found
  books = Book.find_by_solr 'ruby OR splinter', :scores => true
  books.records.first.solr_score
  => 1.21321397
  books.records.last.solr_score
  => 0.12321548

<b>NEW</b>:: Major change on the way the results returned are accessed.
  books = Book.find_by_solr 'ruby'
  # the above will return a SearchResults class with 4 methods:
  # docs|results|records: will return an array of records found
  # 
  #   books.records.is_a?(Array)
  #   => true
  # 
  # total|num_found|total_hits: will return the total number of records found
  # 
  #   books.total
  #   => 2
  # 
  # facets: will return the facets when doing a faceted search
  #
  # max_score|highest_score: returns the highest score found
  # 
  #   books.max_score
  #   => 1.3213213

<b>NEW</b>:: Integrating acts_as_solr to use solr-ruby as the 'backend'. Integration based on the patch submitted by Erik Hatcher
<b>NEW</b>:: Re-factoring rebuild_solr_index to allow adds to be done in batch; and if a finder block is given, it will be called to retrieve the items to index. (thanks Daniel E.)
<b>NEW</b>:: Adding the option to specify the port Solr should start when using rake solr:start 
  rake solr:start RAILS_ENV=your_env PORT=XX

<b>NEW</b>:: Adding deprecation warning for the :background configuration option. It will no longer be updated.
<b>NEW</b>:: Adding support for models that use a primary key other than integer
  class Posting < ActiveRecord::Base
    set_primary_key 'guid' #string
    #make sure you set the :primary_key_field => 'pk_s' if you wish to use a string field as the primary key
    acts_as_solr({},{:primary_key_field => 'pk_s'})
  end

<b>FIX</b>:: Disabling of storing most fields. Storage isn't useful for acts_as_solr in any field other than the pk and id fields. It just takes up space and time. (thanks Daniel E.)
<b>FIX</b>:: Re-factoring code submitted by Daniel E.
<b>NEW</b>:: Adding an :auto_commit option that will only send the commit command to Solr if it is set to true
  class Author < ActiveRecord::Base
	acts_as_solr :auto_commit => false
  end

<b>FIX</b>:: Fixing bug on rake's test task
<b>FIX</b>:: Making acts_as_solr's Post class compatible with Solr 1.2 (thanks Si)
<b>NEW</b>:: Adding Solr 1.2
<b>FIX</b>:: Removing Solr 1.1
<b>NEW</b>:: Adding a conditional :if option to the acts_as_solr call. It behaves the same way ActiveRecord's :if argument option does. 
  class Electronic < ActiveRecord::Base
    acts_as_solr :if => proc{|record| record.is_active?}
  end

<b>NEW</b>:: Adding fixtures to Solr index when using rake db:fixtures:load
<b>FIX</b>:: Fixing boost warning messages
<b>FIX</b>:: Fixing bug when adding a facet to a field that contains boost
<b>NEW</b>:: Deprecating find_with_facet and combining functionality with find_by_solr
<b>NEW</b>:: Adding the option to :exclude_fields when indexing a model
  class User < ActiveRecord::Base
    acts_as_solr :exclude_fields => [:password, :login, :credit_card_number]
  end

<b>FIX</b>:: Fixing branch bug on older ruby version
<b>NEW</b>:: Adding boost support for fields and documents being indexed:
  class Electronic < ActiveRecord::Base
    # You can add boosting on a per-field basis or on the entire document
    acts_as_solr :fields => [{:price => {:boost => 5.0}}], :boost => 5.0
  end

<b>FIX</b>:: Fixed the acts_as_solr limitation to only accept test|development|production environments. 

=== 05-16-2007: Version 0.8.5
<b>FIX</b>:: There's no need to specify the :field_types anymore when doing a search in a model that specifies a field type for a field.
<b>FIX</b>:: Better handling of nil values from indexed fields. Solr complained when indexing fields with field type and the field values being passed as nils.
<b>NEW</b>:: Adding Solr sort (order by) option to the search query (thanks Kevin Hunt)
<b>FIX</b>:: Applying patch suggested for increasing the Solr commit speed (thanks Mourad Hammiche)
<b>FIX</b>:: Updated documentation

=== 05-10-2007: Version 0.8
<b>NEW</b>: New video tutorial
<b>NEW</b>: Faceted search has been implemented and its possible to 'drill-down' on the facets
<b>NEW</b>: New rake tasks you can use to start/stop the solr server in test, development and production environments: (thanks Matt Clark)
  rake solr:start|stop RAILS_ENV=test|development|production (defaults to development if none given)

<b>NEW</b>: Changes to the plugin's test framework and it now supports Sqlite as well (thanks Matt Clark)
<b>FIX</b>: Patch applied (thanks Micah) that allows one to have multiple solr instances in the same servlet
<b>FIX</b>: Patch applied (thanks Micah) that allows indexing of STIs
<b>FIX</b>: Patch applied (thanks Gordon) that allows the plugin to use a table's primary key different than 'id'
<b>FIX</b>: Returning empty array instead of empty strings when no records are found
<b>FIX</b>: Problem with unit tests failing due to order of the tests and speed of the commits

=== 02-16-2007: Version 0.7
<b>NEW</b>: You can now specify the field types when indexing and searching if 
you'd like to preserve its original type:

<b>Indexing</b>


lib/class_methods.rb
require File.dirname(__FILE__) + '/common_methods'
require File.dirname(__FILE__) + '/parser_methods'

module ActsAsSolr #:nodoc:

  module ClassMethods
    include CommonMethods
    include ParserMethods
    
    # Finds instances of a model. Terms are ANDed by default, can be overwritten 
    # by using OR between terms
    # 
    # Here's a sample (untested) code for your controller:
    # 
    #  def search
    #    results = Book.find_by_solr params[:query]
    #  end
    # 
    # You can also search for specific fields by searching for 'field:value'
    # The object returned (ActsAsSolr::SearchResults) works just like a common
    # array and is also compatible with will_paginate helpers.
    # 
    # ====options:
    #
    # will_paginate's compatible pagination
    #
    # page:: - The page you want returned
    # per_page:: - The number of results per page
    # 
    # common pagination
    #
    # offset:: - The first document to be retrieved (offset)
    # limit:: - The number of rows per page
    #
    # conditions:: - An active_record compatible conditions (string, array or hash)
    #                object that is going to be used to further filter the results
    #                returned by this Solr query. You could use this to query
    #                your users but only return the ones that are activated.
    #
    #                User.find_by_solr( 'josh', :conditions => { :activated => true }  )
    #
    # order:: - Orders (sort by) the result set using a given criteria:
    #
    #             Book.find_by_solr 'ruby', :order => 'description asc'
    # 
    # field_types:: This option is deprecated and will be obsolete by version 1.0.
    #               There's no need to specify the :field_types anymore when doing a 
    #               search in a model that specifies a field type for a field. The field 
    #               types are automatically traced back when they're included.
    # 
    #                 class Electronic < ActiveRecord::Base
    #                   acts_as_solr :fields => [{:price => :range_float}]
    #                 end
    # 
    # facets:: This option argument accepts the following arguments:
    #          fields:: The fields to be included in the faceted search (Solr's facet.field)
    #          query:: The queries to be included in the faceted search (Solr's facet.query)
    #          zeros:: Display facets with count of zero. (true|false)
    #          sort:: Sorts the faceted resuls by highest to lowest count. (true|false)
    #          browse:: This is where the 'drill-down' of the facets work. Accepts an array of
    #                   fields in the format "facet_field:term"
    # 
    # Example:
    # 
    #   Electronic.find_by_solr "memory", :facets => {:zeros => false, :sort => true,
    #                                                 :query => ["price:[* TO 200]",
    #                                                            "price:[200 TO 500]",
    #                                                            "price:[500 TO *]"],
    #                                                 :fields => [:category, :manufacturer],
    #                                                 :browse => ["category:Memory","manufacturer:Someone"]}
    # 
    # scores:: If set to true this will return the score as a 'solr_score' attribute
    #          for each one of the instances found. Does not currently work with find_id_by_solr
    # 
    #            books = Book.find_by_solr 'ruby OR splinter', :scores => true
    #            books.records.first.solr_score
    #            => 1.21321397
    #            books.records.last.solr_score
    #            => 0.12321548
    # 
    def find_by_solr(query, options={})
      return [] if query.blank?
      data = parse_query(query.downcase, options)
      return parse_results(data, options) if data
    end
    
    # Finds instances of a model and returns an array with the ids:
    #  Book.find_id_by_solr "rails" => [1,4,7]
    # The options accepted are the same as find_by_solr
    # 
    def find_id_by_solr(query, options={})
      data = parse_query(query, options)
      return parse_results(data, {:format => :ids}) if data
    end
    
    # This method can be used to execute a search across multiple models:
    #   Book.multi_solr_search "Napoleon OR Tom", :models => [Movie]
    # 
    # ====options:
    # Accepts the same options as find_by_solr plus:
    # models:: The additional models you'd like to include in the search
    # results_format:: Specify the format of the results found
    #                  :objects :: Will return an array with the results being objects (default). Example:
    #                               Book.multi_solr_search "Napoleon OR Tom", :models => [Movie], :results_format => :objects
    #                  :ids :: Will return an array with the ids of each entry found. Example:
    #                           Book.multi_solr_search "Napoleon OR Tom", :models => [Movie], :results_format => :ids
    #                           => [{"id" => "Movie:1"},{"id" => Book:1}]
    #                          Where the value of each array is as Model:instance_id
    # 
    def multi_solr_search(query, options = {})
      return nil if query.blank?
      models = "AND (#{solr_configuration[:type_field]}:#{self.name}"
      options[:models].each{|m| models << " OR type_t:"+m.to_s} if options[:models].is_a?(Array)
      options.update(:results_format => :objects) unless options[:results_format]
      data = parse_query(query, options, models<<")")
      result = []
      if data
        docs = data.docs
        return SearchResults.new(:docs => [], :total => 0) if data.total == 0
        if options[:results_format] == :objects
          docs.each{|doc| k = doc.fetch('id').to_s.split(':'); result << k[0].constantize.find_by_id(k[1])}
        elsif options[:results_format] == :ids
          docs.each{|doc| result << {"id"=>doc.values.pop.to_s}}
        end
        SearchResults.new :docs => result, :total => data.total
      end
    end
    
    # returns the total number of documents found in the query specified:
    #  Book.count_by_solr 'rails' => 3
    # 
    def count_by_solr(query, options = {})        
      data = parse_query(query, options)
      data.total_hits
    end
            
    # It's used to rebuild the Solr index for a specific model. 
    #  Book.rebuild_solr_index
    # 
    # If batch_size is greater than 0, adds will be done in batches.
    # NOTE: If using sqlserver, be sure to use a finder with an explicit order.
    # Non-edge versions of rails do not handle pagination correctly for sqlserver
    # without an order clause.
    # 
    # If a finder block is given, it will be called to retrieve the items to index.
    # This can be very useful for things such as updating based on conditions or
    # using eager loading for indexed associations.
    def rebuild_solr_index(batch_size=0, &finder)
      finder ||= lambda { |ar, options| ar.find(:all, options.merge({:order => self.primary_key})) }
    
      if batch_size > 0
        items_processed = 0
        limit = batch_size
        offset = 0
        begin
          items = finder.call(self, {:limit => limit, :offset => offset})
          add_batch = items.collect { |content| content.to_solr_doc }
    
          if items.size > 0
            solr_add add_batch
            solr_commit
          end
    
          items_processed += items.size
          offset += items.size
        end while items.nil? || items.size > 0
      else
        items = finder.call(self, {})
        items.each { |content| content.solr_save }
        items_processed = items.size
      end
      solr_optimize
    end
  end
  
end